src/thor/bidirectional_astar.cc

```diff
@@ -145,70 +133,67 @@ bool BidirectionalAStar::ExpandForward(GraphReader& graphreader,
-                                       offset_time) ||
+    // Expand but only if this isnt the uturn, we'll try that later if nothing else works out
+    disable_uturn = (pred.opp_local_idx() != meta.edge->localedgeidx() &&
+                     ExpandForwardInner(graphreader, pred, nodeinfo, pred_idx, meta, shortcuts, tile,
+                                        offset_time)) ||
                     disable_uturn;
   }
 
   // Handle transitions - expand from the end node of each transition
-  if (!from_transition && nodeinfo->transition_count() > 0) {
+  if (nodeinfo->transition_count() > 0) {
     const NodeTransition* trans = tile->transition(nodeinfo->transition_index());
     for (uint32_t i = 0; i < nodeinfo->transition_count(); ++i, ++trans) {
-      if (trans->up()) {
-        hierarchy_limits_forward_[node.level()].up_transition_count++;
-        disable_uturn = ExpandForward(graphreader, trans->endnode(), pred, pred_idx, true,
-                                      offset_time, invariant) ||
-                        disable_uturn;
-      } else if (!hierarchy_limits_forward_[trans->endnode().level()].StopExpanding()) {
-        disable_uturn = ExpandForward(graphreader, trans->endnode(), pred, pred_idx, true,
-                                      offset_time, invariant) ||
+      // if this is a downward transition (ups are always allowed) AND we are no longer allowed OR
+      // we cant get the tile at that level (local extracts could have this problem) THEN bail
+      graph_tile_ptr trans_tile = nullptr;
+      if ((!trans->up() && hierarchy_limits_forward_[trans->endnode().level()].StopExpanding()) ||
+          !(trans_tile = graphreader.GetGraphTile(trans->endnode()))) {
+        continue;
+      }
+      // setup for expansion at this level
+      hierarchy_limits_forward_[node.level()].up_transition_count += trans->up();
+      const auto* trans_node = trans_tile->node(trans->endnode());
+      EdgeMetadata trans_meta =
+          EdgeMetadata::make(trans->endnode(), trans_node, trans_tile, edgestatus_forward_);
+      uint32_t trans_shortcuts = 0;
+      // expand the edges from this node at this level
+      for (uint32_t i = 0; i < trans_node->edge_count(); ++i, ++trans_meta) {
+        disable_uturn = ExpandForwardInner(graphreader, pred, trans_node, pred_idx, trans_meta,
+                                           trans_shortcuts, trans_tile, offset_time) ||
                         disable_uturn;
       }
     }
   }
 
-  if (!from_transition) {
-    // Now, after having looked at all the edges, including edges on other levels,
-    // we can say if this is a deadend or not, and if so, evaluate the uturn-edge (if it exists)
-    if (!disable_uturn && found_uturn) {
-      // If we found no suitable edge to add, it means we're at a deadend
-      // so lets go back and re-evaluate a potential u-turn
+  // Now, after having looked at all the edges, including edges on other levels,
```

Comment:
this part of the diff looks like a mess but really we had a bunch of branching that was superfluous. instead ive just collapsed it down to only what is needed, a single `if`. the compiler was probably already doing that but honestly this reads way easier anyway